### PR TITLE
Moves and generalizes AmazonAllowedAccountsValidator

### PR DESCRIPTION
**Pull request by ttomsu**
This will enable required group membership validation across all providers, assuming their `credentials` object supports it.

@spinnaker/google-reviewers PTAL

@cfieber @ajordens FYI

### CODE REVIEW

File: clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  void validate(String user, Collection<String> allowedAccounts, Object description, Errors errors) {
+    if (!accountCredentialsProvider.all.find { it.requiredGroupMembership }) {
+      // no accounts have group restrictions so no need to validate / log
+      return
+    }
+
+    def json = null
+    try {
+      json = OBJECT_MAPPER.writeValueAsString(description)
+    } catch (Exception ignored) {
```
Comment (lwander) [inline clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy]: What's going on here?

File: clouddriver-security/src/test/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidatorSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    when:
+    def description = new TestDescription(credentials: credentialsWithRequiredGroup)
+
+    validator.validate("TestUser", allowedAccounts, description, errors)
+
+    then:
+    rejectValueCount * errors.rejectValue("credentials", "unauthorized", _)
+
+    where:
+    allowedAccounts                      || rejectValueCount
```
Comment (lwander) [inline clouddriver-security/src/test/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidatorSpec.groovy]: This syntax seems off, I think you only only need a single `|`

**Comment by lwander** [comment]
LGTM post comments

File: clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy (hunk 1 above)
Comment (ttomsu) [inline clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy]: Looks like it's used only for debugging. This was migrated from the other version.

File: clouddriver-security/src/test/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidatorSpec.groovy (hunk 2 above)
Comment (ttomsu) [inline clouddriver-security/src/test/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidatorSpec.groovy]: It's legit -https://spockframework.github.io/spock/docs/1.0/data_driven_testing.html#_syntactic_variations

Comment (ajordens) [inline clouddriver-security/src/test/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidatorSpec.groovy]: Should use `||` when separating inputs from expected outputs.

File: clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy (hunk 1 above)
Comment (ajordens) [inline clouddriver-security/src/main/groovy/com/netflix/spinnaker/clouddriver/security/DefaultAllowedAccountsValidator.groovy]: You changed the copyright ... you own it now!

**Comment by ajordens** [comment]
From our earlier conversation ... I don't imagine this is going to be a useful class long term with the move to per-application and per-account read/write permissions (_split out into a separate service_).

That said, if there's some short-term benefit for locking down write-access than great!

**Comment by ttomsu** [comment]
It's a step forward in the right direction, which is why I generalized this class. Once everything in Gate is merged and released, we can finally have a good, even if not yet complete, story how to lock down access.

**Pull request merged**